### ISSUE TITLE
we need our own input thread

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#2136]
See #2131 for some motivation regarding the Kitty protocol, which inhibits `SIGINT` and friends. We're at the point where we need our own input thread. This would simplify a lot of code, especially cursor location report handling (we'd no longer need the dubious `cursor_location_simple()`). It's basically necessary for anything where *we* need respond to input rather than the client. This includes at least synthesizing signals (which ought only be done if line discipline signals have not been disabled) and cursor location reports.

This would be a big change for 2.4.0, and maybe it's better for 3.0.0? But right now we have #2131 scheduled for 2.4.0, and we need this to accomplish #2131...alright, bumping #2131 back to 3.0.0.

**Comment by dankamongmen** [comment]
so what's this thread doing?

it's got two output queues, each with a mutex and condvar: one of cursor location reports (and other future messages to us from the terminal), and one of processed user input (`ncinput`s). we're constantly poll()ing on either one or two descriptors (on Windows, we're using `WaitForMultipleObjects()` on the input handle). whenever there's data available, we read it in a large burst.

all lexing state must be in a self-contained struct carried across reads, so that we can handle mutlibyte sequences (escapes and/or UTF-8) that cross a read boundary.

at each lexing segmentation (end of a UTF8 character or escape), add to the appropriate queue, and signal its condvar.

if an output queue is full, drop its source from the poll. keep looking at any other sources. otherwise, block on a condvar. we'll need a pipe for each queue as well, and we'll want to poll on *that*.

on read, lock the queue. extract the next item if there are any items, and unlock. otherwise, block on a condvar. if the queue was full when you read it, signal the general condvar and write to your pipe.

when we want a cursor report now, we can just freely block on that queue after firing off a `u7`.

**Comment by dankamongmen** [comment]
started this work in `dankamongmen/input-thread`

**Comment by dankamongmen** [comment]
it would be really nice if, as a result of this, xterm could properly, responsively switch blitters in a `ncplayer` video instance. this has always been problematic.

**Comment by dankamongmen** [comment]
so one worry here is the case where we are *not* redirected from a file, but *do* collect "unbounded" input without flushing it. we should probably accept a `FILE*` in `notcurses_core_init()` for input, just as we do output, and it could be `NULL` to indicate "i don't care about input". that breaks ABI, though, and `NULL` for the other argument means "use stdout", so it would be weird for a second `NULL` to mean "use nothing". the caller could close stdin and open `/dev/null` (though probably not on windows, and this is a terrible solution anyway, listed only for completeness). we could accept a new option, `NCOPTION_DRAIN_INPUT`, indicating that input may be freely thrown away. we could require all callers to actively handle input.

i think the third of those ideas is the best. if you're not handling input, you ought provide `NCOPTION_DRAIN_INPUT`. in this case, we can toss input when stdin is connected to a terminal, and thus always be able to read escapes.

**Comment by dankamongmen** [comment]
if `NCOPTION_DRAIN_INPUT` is provided, and `stdin` is not connected to the terminal, we ought just never read from it at all, since we're never going to use any of the input. no reason to read it just to drop it.

**Comment by dankamongmen** [comment]
so this still leaves us with the question of how to handle initial query responses....do we have them go through the new input layer, or one-off them? if we want to do the former, we need expand the control sequences queue to encompass more than one type of control (probably a good thing anyway). we'd still need all the automaton from the existing code, just folded into the general control-processing loop.

if we don't do it that way, we lose all the robustness and correctness improvements we're gaining from this work, and go back to a world where someone can press a key in the middle of interrogation, and we lose that input.

so i think the former is the way to do it, and just pop off control sequences until we hit the DA1 response. or the second cursor location request. hrmmmm, if cursor location reports are as widely supported as DA1, we never need actually send DA1, do we?

either way this is gonna take some work. i think we ought cut 2.4.1 now.

**Comment by dankamongmen** [comment]
ok lots of good progress. terminal interrogation now blocks on processing of initial responses, which we are reading. we need copy over our previous automaton. ideally, we'd unite it with the mainline CSI automaton (regular languages are closed over union so it's of course possible). we'll set `->complete` and signal the condvar when we read the response to DA1. this ought reduce startup time latency because we'll now be reading responses and doing setupterm crap in parallel, huzzah.

**Comment by dankamongmen** [comment]
further progress. we now lex up the initial responses in our new input thread, and block the terminal interrogation on it. we're not yet using the collected data in interrogation, need to do that. we'll also need:

* [ ] guard against `initdata` accesses once it's gone `NULL`
* [ ] put cursor location reports on that queue
* [ ] parse up bulk inputs

**Comment by dankamongmen** [comment]
ok, cursor location reports are now complete, and much better than they ever were before. i've removed `input.h` from my branch.

**Comment by dankamongmen** [comment]
closed the client bulk input loop. now we just need get back to the level of the old code in terms of multibyte UTF8, mouse inputs, and stats, and we'll be done =]

**Comment by dankamongmen** [comment]
we're pretty much there now. we appear to deadlock on a redirected input from large files, so we need to resolve that. this likely gets merged today.

**Comment by dankamongmen** [comment]
we also need load the special characters, and extract modifiers, both for non-kitty setups.

**Comment by dankamongmen** [comment]
merged! ok everybody let's test the hell out of this
